### PR TITLE
Bessel function of higher order

### PR DESCRIPTION
**Pull request by Yaraslaut**
This PR adds Bessel functions of higher order (#5637)

TODO:
- [x] update unit test so they cover all the code and ckeck precision for higher orders as well

### CODE REVIEW

**Comment by PhilMiller** [comment]
If you end up rebasing for whatever reason, there's a typo in the first commit message, 'fucntions'

**Comment by PhilMiller** [comment]
OK to test

**Comment by Yaraslaut** [comment]
> If you end up rebasing for whatever reason, there's a typo in the first commit message, 'fucntions'

oh, will then ask to squash pr to omit typos

**Comment by mhoemmen** [comment]
I'm not a Kokkos dev, but would you consider calling the existing mathematical special functions where appropriate?

* Kokkos requires C++17, which adds [mathematical special functions](https://en.cppreference.com/w/cpp/numeric/special_functions)
* CUDA has [device implementations](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH.html#group__CUDA__MATH) of the mathematical special functions

Not using existing implementations where possible risks pessimizing both performance and accuracy.

**Comment by Yaraslaut** [comment]
> I'm not a Kokkos dev, but would you consider calling the existing mathematical special functions where appropriate?
> 
>     * Kokkos requires C++17, which adds [mathematical special functions](https://en.cppreference.com/w/cpp/numeric/special_functions)
> 
>     * CUDA has [device implementations](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH.html#group__CUDA__MATH) of the mathematical special functions
> 
> 
> Not using existing implementations where possible risks pessimizing both performance and accuracy.

Hi, here is small example of how this functions can be called
```
// adl
void t()
{
  using namespace std;
  using namespace Kokkos::Experimental;

  auto d{2.0};
  cyl_bessel_j(2,d);

  auto std_complex = std::complex<double>{2.0,2.0};
  //cyl_bessel_j(2,std_complex); // error

  auto kokkos_complex = Kokkos::complex<double>{2.0,2.0};
  //cyl_bessel_j(2,kokkos_complex); //error
  cyl_bessel_j<decltype(kokkos_complex),double,int>(2,kokkos_complex);
}


int main(int argc, char* argv[]) {
  Kokkos::ScopeGuard guard(argc, argv);
  auto d = 2.0;
  auto kokkos_complex = Kokkos::complex<double>(2.0,2.0);
  auto std_complex = std::complex<double>(2.0,2.0);
  std::cyl_bessel_i(1, d);
  //std::cyl_bessel_i(1, kokkos_complex);  error
  //std::cyl_bessel_i(1, std_complex); error

  //Kokkos::Experimental::cyl_bessel_j1<decltype(d),double,int>(d); //error
  Kokkos::Experimental::cyl_bessel_j1<decltype(kokkos_complex),double,int>(kokkos_complex);
  //Kokkos::Experimental::cyl_bessel_j1<decltype(std_complex),double,int>(std_complex); //error
}


```
 * bessel functions from kokkos require explicit template arguments since they can not deduce all types from one argument only. 
 * functions from std have real arguments only so if user overlap namespaces and call bessel function directly without namespace specification it will end up calling std functoin ( `void t()` function )
  * kokkos functions can take as argument only `Kokkos::complex<>` type 
  
I think it is a good idea to have bessel functions for real types in Kokkos namespace so users can just call `Kokkos::cyl_besselj(2,2.0)` but that another topic of how properly expose special functions from `<cmath>` in Kokkos namespace so ADL can call functions

**Comment by fnrizzi** [comment]
@Yaraslaut just started looking at this. 
First comment i have: 
```
auto kokkos_complex = Kokkos::complex<double>{2.0,2.0};
cyl_bessel_j<decltype(kokkos_complex),double,int>(2,kokkos_complex);
```  
I think it would be useful to allow doing this: 
```
cyl_bessel_j(2,kokkos_complex);

**Comment by Yaraslaut** [comment]
> @Yaraslaut I just started looking at this, thanks for the PR !
> 
>     1. first comment:
> 
> 
> ```c++
> auto kokkos_complex = Kokkos::complex<double>{2.0,2.0};
> cyl_bessel_j<decltype(kokkos_complex),double,int>(2,kokkos_complex);
> ```
> 
> I think it would be useful to allow doing this:
> 
> ```c++
> auto kokkos_complex = Kokkos::complex<double>{2.0,2.0};
> cyl_bessel_j(2,kokkos_complex); 
> ```
> 
> where this could be handled as:
> 
> ```c++
> template <class RealType, class IntType>
> KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType a,
>                                               const Kokkos::complex<RealType> & z,
>                                               const RealType& joint_val = 25,
>                                               const IntType& bw_start   = 70) {
> ```
> 
>     2. second comment: I agree with @mhoemmen: ideally we can expose as rich of an API as needed (for example, also to allow users to avoid having to specify all templates like in my example above), but internally we should call native libraries if available and if nothing is avaible have a fallback implementation (effectively what you have in this PR could be that)

Thanks for reviewing this PR, I will try to address this issues, and let you know once it is ready for further review.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by Rombur** [comment]
The new tests fail with ROCm 5.6 but there is a known issue with ROCm 5.5 and 5.6. If the testsuite passes for ROCm 5.2, I am OK  with disabling the new tests for both ROCm 5.5 and 5.6.

**Comment by crtrott** [comment]
@Rombur can you do that fix for turning off the test for ROCM 5.5/5.6?

**Comment by Rombur** [comment]
Sure

**Comment by Rombur** [comment]
@Yaraslaut Your latest commit uses C++ math functions but they are not implemented in libc++ see [here](https://en.cppreference.com/w/cpp/compiler_support/17)

**Comment by Yaraslaut** [comment]
> @Yaraslaut Your latest commit uses C++ math functions but they are not implemented in libc++ see [here](https://en.cppreference.com/w/cpp/compiler_support/17)

thanks for pointing this out, hopefully I fixed it.

**Comment by dalg24** [comment]
> It fixes that error but there are others:
> 
> ```
> /var/jenkins/workspace/Kokkos/install/include/Kokkos_MathematicalSpecialFunctions.hpp(1328): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> /var/jenkins/workspace/Kokkos/install/include/Kokkos_Complex.hpp(645): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> /var/jenkins/workspace/Kokkos/install/include/Kokkos_Complex.hpp(599): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> /var/jenkins/workspace/Kokkos/install/include/Kokkos_Complex.hpp(645): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> ```
> 
> ```
> /var/jenkins/workspace/Kokkos/core/src/Kokkos_MathematicalSpecialFunctions.hpp(1328): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> /var/jenkins/workspace/Kokkos/core/src/Kokkos_MathematicalSpecialFunctions.hpp(1343): error: identifier "Kokkos::numbers::pi_v<double> " is undefined in device code
> ```

The issue is that 
https://github.com/kokkos/kokkos/blob/41cf2e51cd8e63a614aa1b4a9564a35934b805d5/core/src/Kokkos_Complex.hpp#L643-L647 takes the scalar argument by reference and then (same problem) calls the complex constructor from the real part that also takes it by reference
https://github.com/kokkos/kokkos/blob/41cf2e51cd8e63a614aa1b4a9564a35934b805d5/core/src/Kokkos_Complex.hpp#L114
and, at the call site you did
```C++
 CmplxType factor = Kokkos::exp(CmplxType(0.0, 1.0) *
                                Kokkos::numbers::pi_v<RealType> * pos_a);
```
where you see the complex `CmplxType(0.0, 1.0) ` multiplies scalar `Kokkos::numbers::pi_v<RealType>` while we unfortunately cannot take the address of the Kokkos math constants on the device side.  This is a known issue:
https://github.com/kokkos/kokkos/blob/41cf2e51cd8e63a614aa1b4a9564a35934b805d5/core/unit_test/TestMathematicalConstants.hpp#L64-L72

We need to document that pitfall.

The way to work around this is to declare a local variable and make a copy.
```C++
auto pi =  Kokkos::numbers::pi_v<RealType>;
CmplxType factor = Kokkos::exp(CmplxType(0.0, 1.0) * pi * pos_a);
```

**Comment by Yaraslaut** [comment]
Thanks @dalg24 fot the help in fixing this odd error, I pushed the fix

**Comment by masterleinad** [comment]
Looks like the the test fails when using `std` functions:
```
4: [ RUN      ] hip.mathspecialfunc_cbesselj2y2
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2132: Failure
4: Expected: (Kokkos::abs(h_cbj2(i) - h_ref_cbj2(i))) <= (Kokkos::abs(h_ref_cbj2(i)) * 1e-13), actual: 0.947984 vs 1.22779e-13
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2132: Failure
4: Expected: (Kokkos::abs(h_cbj2(i) - h_ref_cbj2(i))) <= (Kokkos::abs(h_ref_cbj2(i)) * 1e-13), actual: 0.947984 vs 1.22779e-13
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2132: Failure
4: Expected: (Kokkos::abs(h_cbj2(i) - h_ref_cbj2(i))) <= (Kokkos::abs(h_ref_cbj2(i)) * 1e-13), actual: 0.947984 vs 1.22779e-13
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2132: Failure
4: Expected: (Kokkos::abs(h_cbj2(i) - h_ref_cbj2(i))) <= (Kokkos::abs(h_ref_cbj2(i)) * 1e-13), actual: 0.947984 vs 1.22779e-13
4: [  FAILED  ] hip.mathspecialfunc_cbesselj2y2 (0 ms)
4: [ RUN      ] hip.mathspecialfunc_cbesseli2k2
4: [       OK ] hip.mathspecialfunc_cbesseli2k2 (0 ms)
4: [ RUN      ] hip.mathspecialfunc_cbesselh12h22
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.432749 vs 2.41179e-13
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.432749 vs 2.50098e-13
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.226039 vs 5.11872e-14
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.226039 vs 5.11872e-14
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.00343646 vs 1.66666e-14
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2491: Failure
4: Expected: (Kokkos::abs(h_cbh12(i) - h_ref_cbh12(i))) <= (Kokkos::abs(h_ref_cbh12(i)) * 1e-13), actual: 0.00343646 vs 1.66666e-14
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 0.182976 vs 7.73567e-15
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 0.182976 vs 7.73567e-15
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.38258e-05 vs 7.45952e-19
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.38258e-05 vs 7.45952e-19
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.81657e-05 vs 6.79118e-19
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.81657e-05 vs 6.79118e-19
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.44286e-05 vs 4.66927e-19
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestMathematicalSpecialFunctions.hpp:2497: Failure
4: Expected: (Kokkos::abs(h_cbh22(i) - h_ref_cbh22(i))) <= (Kokkos::abs(h_ref_cbh22(i)) * 1e-13), actual: 1.44286e-05 vs 4.66927e-19
4: [  FAILED  ] hip.mathspecialfunc_cbesselh12h22 (1 ms)
```

**Comment by Rombur** [comment]
@masterleinad No I don't think so. I think that's the compiler bug that we discussed before in this PR. We just need to disable the failing tests for ROCm 5.5 and 5.6

**Comment by masterleinad** [comment]
> @masterleinad No I don't think so. I think that's the compiler bug that we discussed earlier in this PR. We just need to disable the failing tests for ROCm 5.5 and 5.6

Even better!

**Comment by Yaraslaut** [comment]
I think that PR is somewhat ready for review, sorry it took so long to get it working after initial implementation it :(
not sure whom should i ping to review this changes, @fnrizzi ? @Rombur ?

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 1]
```diff
@@ -1268,6 +1268,222 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+  }
+
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25.0,
+                                              const IntType& bw_start   = 70) {
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1293]: I would rather not have these extra arguments (and remove the from the current public interface as well). These are implementation details.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 2]
```diff
@@ -1268,6 +1268,222 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(0.0, 0.0);
+  }
+  auto pos_a  = Kokkos::abs(a);
+  auto factor = a < 0 ? Kokkos::pow(-1.0, pos_a) : 1.0;
+  return factor *
+         cyl_bessel<cyl_bessel_j0<CmplxType, RealType, IntType>,
+                    cyl_bessel_j1<CmplxType, RealType, IntType>, CmplxType,
+                    IntType, RealType>(pos_a, z, joint_val, bw_start);
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1302]: Can't we just pass the values for `cyl_bessel_j0` and `cyl_bessel_j1` on? (Not necessarily a problem).

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 3]
```diff
@@ -694,51 +694,61 @@ struct TestComplexBesselJ0Y0Function {
...
+        Kokkos::Experimental::Impl::cyl_bessel_y0<Kokkos::complex<double>,
+                                                  double, int>(d_z(i));
   }
 
   KOKKOS_INLINE_FUNCTION
   void operator()(const TestLargeArgTag&, const int& /*i*/) const {
     d_cbj0_large(0) =
-        Kokkos::Experimental::cyl_bessel_j0<Kokkos::complex<double>, double,
-                                            int>(d_z_large(0));
+        Kokkos::Experimental::Impl::cyl_bessel_j0<Kokkos::complex<double>,
```
Comment (masterleinad) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:708]: Why did you move `cyl_bessel_*[0|1]` to `Impl`? If we really want to remove them from the public interface, we should rather deprecate them (in the public interface).

**Review by masterleinad — COMMENTED** [review]
Not a complete review. Just some obeservations.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp (hunk 2 above)
Comment (Yaraslaut) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:1302]: previous implementation was using recursion so this functions where required
Not sure that we can change it now since it will require constexpr evaluation of `cyl_ besse{}{} ` and if we pass them by value then for zero and first order instead of one evaluation we end up with two

File: core/unit_test/TestMathematicalSpecialFunctions.hpp (hunk 3 above)
Comment (Yaraslaut) [reply core/unit_test/TestMathematicalSpecialFunctions.hpp:708]: I moved it there so user will not call  `cyl_bessel_*[0|1]`since native functions can exist for cuda or std, and it creates more uniform interface 
if exist better solution to hide this functions please tell me

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:1302]: I think it's fine.

**Comment by masterleinad** [comment]
We still need to disable the tests failing for `ROCm` 5.5 and 5.6.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 4]
```diff
@@ -1268,6 +1268,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+  }
+
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25.0,
+                                              const IntType& bw_start   = 70) {
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1293]: In my opinion, these template arguments should only be exposed where they are directly used, i.e., in `cyl_bessel_j0` and `cyl_bessel_j1`.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 5]
```diff
@@ -1268,6 +1268,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+         cyl_bessel<cyl_bessel_j0<CmplxType, RealType, IntType>,
+                    cyl_bessel_j1<CmplxType, RealType, IntType>, CmplxType,
+                    IntType, RealType>(pos_a, z, joint_val, bw_start);
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_y(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25,
+                                              const IntType& bw_start   = 70) {
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1309]: Similarly here.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 6]
```diff
@@ -1268,6 +1268,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+  return factor *
+         cyl_bessel<cyl_bessel_h20<CmplxType>, cyl_bessel_h21<CmplxType>,
+                    CmplxType, IntType>(pos_a, z);
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_i(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25,
+                                              const IntType& bw_start   = 70) {
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1354]: and here

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 7]
```diff
@@ -1268,6 +1268,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+    bessel_im2    = tmp;
+  }
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_k(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 9,
+                                              const IntType& bw_start   = 30) {
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1374]: and here.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp (hunk 4 above)
Comment (Yaraslaut) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:1293]: it is impossible to omit this parameters here, since templates and default arguments does not play nicely together, see:  https://godbolt.org/z/o4fG5d69f

Comment (masterleinad) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:1293]: Right. I suggest specifying the necessary arguments in this function explicitly instead of exposing them as function parameters.

Comment (Yaraslaut) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp]: I updated PR

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 8]
```diff
@@ -694,51 +694,61 @@ struct TestComplexBesselJ0Y0Function {
 
   KOKKOS_INLINE_FUNCTION
   void operator()(const int& i) const {
-    d_cbj0(i) = Kokkos::Experimental::cyl_bessel_j0<Kokkos::complex<double>,
-                                                    double, int>(d_z(i));
-    d_cby0(i) = Kokkos::Experimental::cyl_bessel_y0<Kokkos::complex<double>,
-                                                    double, int>(d_z(i));
+    d_cbj0(i) =
+        Kokkos::Experimental::Impl::cyl_bessel_j0<Kokkos::complex<double>,
```
Comment (masterleinad) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:698]: Can you please change all of these to use/test the public interface?

**Review by masterleinad — COMMENTED** [review]
Apart from testing the public interface instead of the function in `Impl`, this looks good to me (pending CI).

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 9]
```diff
@@ -1901,8 +1933,653 @@ struct TestComplexBesselH2Function {
...
+    d_cbj9(0) = Kokkos::Experimental::cyl_bessel_j(9, d_z(0));
+    d_cbj9(1) = Kokkos::Experimental::cyl_bessel_j(9, d_z(1));
+    d_cbj9(2) = Kokkos::Experimental::cyl_bessel_j(9, d_z(2));
+    d_cbj9(3) = Kokkos::Experimental::cyl_bessel_j(9, d_z(3));
+    d_cbj9(4) = Kokkos::Experimental::cyl_bessel_j(9, d_z(4));
+  }
+};
+
+template <class ExecSpace>
+struct TestBesselCalsWithDeduction {
```
Comment (fnrizzi) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:2558]: is `Cals` a typo? is it supposed to be `Calls`?

File: core/unit_test/TestMathematicalSpecialFunctions.hpp [hunk 10]
```diff
@@ -1901,8 +1933,653 @@ struct TestComplexBesselH2Function {
...
+    Kokkos::Experimental::cyl_bessel_h1(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_h2(6, Kokkos::complex<double>(1.0, 1.0));
+
+    // Call to vendor functions of possible and ret real type
+    Kokkos::Experimental::cyl_bessel_j(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_i(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_k(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_y(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_h1(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_h2(6, 1.0);
```
Comment (fnrizzi) [inline core/unit_test/TestMathematicalSpecialFunctions.hpp:2582]: what is being tested here? just that it compiles?

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 11]
```diff
@@ -1268,6 +1269,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+#if defined(__cpp_lib_math_special_functions)
+  KOKKOS_IF_ON_HOST((return std::cyl_bessel_k(a, z);))
+#endif
+  return real(cyl_bessel_k(a, Kokkos::complex<RealType>(z)));
+}
+
+template <typename RealType, typename IntType>
+KOKKOS_INLINE_FUNCTION RealType cyl_bessel_y(const IntType& a,
+                                             const RealType& z) {
+  return real(cyl_bessel_y(a, Kokkos::complex<RealType>(z)));
```
Comment (fnrizzi) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1465]: why here we are not calling the cuda one like you did for `_j`?
I think cuda supports this too: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__DOUBLE.html
or at least the version of the toolkit shown that

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 12]
```diff
@@ -1268,6 +1269,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z) {
+  constexpr RealType joint_val = 25.0;
+  constexpr IntType bw_start   = 70;
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(0.0, 0.0);
+  }
+  auto pos_a  = Kokkos::abs(a);
+  auto factor = a < 0 ? Kokkos::pow(-1.0, pos_a) : 1.0;
```
Comment (fnrizzi) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1299]: pos_a and factor can both be marked as const

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 13]
```diff
@@ -1268,6 +1269,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_y(const IntType& a,
+                                              const CmplxType& z) {
+  constexpr RealType joint_val = 25.0;
+  constexpr IntType bw_start   = 70;
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(-infinity<double>::value, 0.0);
+  }
+  auto pos_a  = Kokkos::abs(a);
+  auto factor = a < 0 ? Kokkos::pow(-1.0, pos_a) : 1.0;
```
Comment (fnrizzi) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1315]: pos_a and factor can both be marked as const

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 14]
```diff
@@ -1268,6 +1269,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h1(const IntType& a,
+                                               const CmplxType& z) {
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(0.0, 0.0);
+  }
+  auto pos_a       = Kokkos::abs(a);
+  auto pi          = Kokkos::numbers::pi_v<RealType>;
+  CmplxType factor = Kokkos::exp(CmplxType(0.0, 1.0) * pi * pos_a);
```
Comment (fnrizzi) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1330]: pos_a and factor and pi can all be marked as const

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp [hunk 15]
```diff
@@ -1268,6 +1269,304 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
...
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h2(const IntType& a,
+                                               const CmplxType& z) {
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(0.0, 0.0);
+  }
+
+  auto pos_a       = Kokkos::abs(a);
+  auto pi          = Kokkos::numbers::pi_v<RealType>;
+  CmplxType factor = Kokkos::exp(-CmplxType(0.0, 1.0) * pi * pos_a);
```
Comment (fnrizzi) [inline core/src/Kokkos_MathematicalSpecialFunctions.hpp:1345]: pos_a and factor and pi can all be marked as const

**Comment by masterleinad** [comment]
You'll have to fix the indentation.

**Comment by Yaraslaut** [comment]
> You'll have to fix the indentation.

Thanks, for pointing this our,otherwise i would discover it only after a while. I tried to addres all issues, still not sure what to do with large arguments since they require customization of  default variables to get proper result

File: core/unit_test/TestMathematicalSpecialFunctions.hpp (hunk 10 above)
Comment (Yaraslaut) [reply core/unit_test/TestMathematicalSpecialFunctions.hpp:2566]: Yes, only that it properly deducing templates, i can also add tests for the numerical value of those

Comment (masterleinad) [reply core/unit_test/TestMathematicalSpecialFunctions.hpp:2566]: Please do.

File: core/src/Kokkos_MathematicalSpecialFunctions.hpp (hunk 11 above)
Comment (Yaraslaut) [reply core/src/Kokkos_MathematicalSpecialFunctions.hpp:1477]: Thanks for spotting this issue, I added it there

File: core/unit_test/TestMathematicalSpecialFunctions.hpp (hunk 10 above)
Comment (Yaraslaut) [reply core/unit_test/TestMathematicalSpecialFunctions.hpp]: I added tests for higher orders for both real and complex arguments, and I was lucky to find out divergene between results taken from octave and julia. So I need to rethink implementation, if anyone have any ideas i will be happy to hear them

**Comment by fnrizzi** [comment]
making this a draft following the comment: https://github.com/kokkos/kokkos/pull/5897#discussion_r1378422127

**Comment by Yaraslaut** [comment]
Search for an existing implementation led me to discover implementation of Bessel function by Amos[1].  You can see  this is a Fortran code and i am not sure if kokkos wants to link/compile this library, perhaps @fnrizzi  or @masterleinad  can advice  how we can proceed with this PR.  


[1] https://netlib.org/amos/ 
Few other references for the history
[2] https://github.com/JuliaMath/openspecfun/tree/main/amos
[3] COMPUTATION OF BESSEL FUNCTIONS OF COMPLEX ARGUMENT BY D. E. AMOS, SAND83-0083, MAY, 1983.
[4] COMPUTATION OF BESSEL FUNCTIONS OF COMPLEX ARGUMENT AND LARGE ORDER BY D. E. AMOS, SAND83-0643, MAY, 1983
[5] SUBROUTINE PACKAGE FOR BESSEL FUNCTIONS OF A COMPLEX ARGUMENT AND NONNEGATIVE ORDER BY D. E. AMOS, SAND85-1018, MAY, 1985
[6] PORTABLE PACKAGE FOR BESSEL FUNCTIONS OF A COMPLEX ARGUMENT AND NONNEGATIVE ORDER BY D. E. AMOS, TRANS. MATH. SOFTWARE, 1986

**Comment by masterleinad** [comment]
> Search for an existing implementation led me to discover implementation of Bessel function by Amos[1]. You can see this is a Fortran code and i am not sure if kokkos wants to link/compile this library, perhaps @fnrizzi or @masterleinad can advice how we can proceed with this PR.

I don't think we want to use an external library just to be able to use Bessel functions (and does that even work with the GPU backends?). I would prefer just adopting a suitable implementation and converting it to `Kokkos` code (adding the appropriate copyrights, of course).

**Comment by fnrizzi** [comment]
> > Search for an existing implementation led me to discover implementation of Bessel function by Amos[1]. You can see this is a Fortran code and i am not sure if kokkos wants to link/compile this library, perhaps @fnrizzi or @masterleinad can advice how we can proceed with this PR.
> 
> I don't think we want to use an external library just to be able to use Bessel functions (and does that even work with the GPU backends?). I would prefer just adopting a suitable implementation and converting it to `Kokkos` code (adding the appropriate copyrights, of course).

i agree with Daniel!

**Comment by Yaraslaut** [comment]
@masterleinad  @fnrizzi existing libraries does not support complex arguments, for example see boost: https://github.com/boostorg/math/issues/689
